File: Haskell/13-most-frequent-symbol.hs
import Data.Array (Array, listArray, (!))
import Control.Monad (replicateM)
import Data.List (maximumBy)
import Data.Ord (comparing)


data Query = Query { left :: Int, right :: Int }

data SegmentTree =
    Leaf {ix :: Int, freqs :: [Int]} |
    Node {ix :: Int
    , leftEnd :: Int
    , rightEnd :: Int
    , leftChild :: SegmentTree
    , rightChild :: SegmentTree
    , freqs :: [Int]} deriving Show


alphabet :: [Char]
alphabet = ['a'..'z']


alphabetSize :: Int
alphabetSize = length alphabet


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readQueries :: Int -> IO [Query]
readQueries nrQueries = do
    lines' <- replicateM nrQueries getLine
    return $ map (\line -> let [a, b] = convertToIntList line in Query a b) lines'


findChar :: [Int] -> Char
findChar frequencies = fst $ maximumBy (comparing snd) $ zip alphabet frequencies


combine :: [Int] -> [Int] -> [Int]
combine frequencies1 frequencies2 = map (\(f1, f2) -> f1 + f2) $ zip frequencies1 frequencies2


createSingleLetterFrequencyList :: Char -> [Int]
createSingleLetterFrequencyList char = map (\c -> if c == char then 1 else 0) alphabet


calcSegmentTree :: Array Int Char -> Int -> SegmentTree
calcSegmentTree chars n =
    let buildTree :: Int -> Int -> Int -> SegmentTree
        buildTree ix left right
            | left == right = Leaf left $ createSingleLetterFrequencyList $ chars ! left
            | otherwise =
                let ix' = 2 * ix + 1
                    ix'' = ix' + 1
                    middleIx = (left + right) `div` 2
                    leftChild = buildTree ix' left middleIx
                    rightChild = buildTree ix'' (middleIx + 1) right
                in Node ix left right leftChild rightChild $ combine (freqs leftChild) (freqs rightChild)
        in buildTree 0 0 (n - 1)


mostFrequentCharsInRange :: Array Int Char -> [Query] -> Int -> [Char]
mostFrequentCharsInRange chars queries n =
    let segmentTree = calcSegmentTree chars n
        getFrequenciesInSegment :: SegmentTree -> Int -> Int -> Int -> Int -> [Int]
        getFrequenciesInSegment tree leftEnd rightEnd i j
            | i <= leftEnd && j >= rightEnd = freqs tree
            | i > rightEnd || j < leftEnd = replicate n 0
            | otherwise =
                let middleIx = (leftEnd + rightEnd) `div` 2
                    frequencies1 = getFrequenciesInSegment (leftChild tree)  leftEnd middleIx i j
                    frequencies2 = getFrequenciesInSegment (rightChild tree) (middleIx + 1) rightEnd i j
                in combine frequencies1 frequencies2
    in map (\(Query a b) ->
        let freqs = getFrequenciesInSegment segmentTree 0 (n - 1) (a - 1) (b - 1) in findChar freqs) queries


main :: IO()
main = do
    string <- getLine
    let n = length string
    let chars = listArray (0, n - 1) string
    nrQueries <- readLn
    queries <- readQueries nrQueries
    let result = mostFrequentCharsInRange chars queries n
    mapM_ putStrLn $ map (: []) result

File: Python/13_most_frequent_symbol.py
import sys

from math import ceil, log2


ALPHABET = 'abcdefghijklmnopqrstuvwxyz'
SIZE = len(ALPHABET)


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def read_queries(reader, n):
    queries = []
    for _ in range(n):
        a, b = convert_to_intlist(next(reader))
        queries.append((a, b))
    return queries


def find_character(frequencies):
    char, _ = max(zip(ALPHABET, frequencies), key=lambda x: x[1])
    return char


def combine(frequencies1, frequencies2):
    return tuple(f1 + f2 for f1, f2 in zip(frequencies1, frequencies2))


def create_single_letter_frequency_list(char):
    return tuple(1 if letter == char else 0 for letter in ALPHABET)


def calc_segment_tree(array, n):
    max_nr_nodes = 2 ** (ceil(log2(n)) + 1) - 1
    default_frequencies = (0,) * SIZE
    segment_tree = [default_frequencies] * max_nr_nodes

    def build_tree(node_ix, left_end, right_end):
        if left_end == right_end:
            segment_tree[node_ix] = create_single_letter_frequency_list(array[left_end])
        else:
            left_child_ix = 2 * node_ix + 1
            right_child_ix = left_child_ix + 1
            middle_ix = (left_end + right_end) // 2
            build_tree(left_child_ix, left_end, middle_ix)
            build_tree(right_child_ix, middle_ix + 1, right_end)
            segment_tree[node_ix] = combine(segment_tree[left_child_ix], segment_tree[right_child_ix])

    build_tree(0, 0, n - 1)
    return segment_tree


def most_frequent_chars_in_range(array, queries):
    n = len(array)
    segment_tree = calc_segment_tree(array, n)

    def get_most_frequent_char_in_segment(node_ix, left_end, right_end, i, j):
        if i <= left_end and j >= right_end:
            return segment_tree[node_ix]
        if i > right_end or j < left_end:
            return (0,) * SIZE
        middle_ix = (left_end + right_end) // 2
        frequencies1 = get_most_frequent_char_in_segment(2*node_ix+1, left_end, middle_ix, i, j)
        frequencies2 = get_most_frequent_char_in_segment(2*node_ix+2, middle_ix + 1, right_end, i, j)
        return combine(frequencies1, frequencies2)

    result = []
    for a, b in queries:
        freqs = get_most_frequent_char_in_segment(0, 0, n - 1, a - 1, b - 1)
        result.append(find_character(freqs))
    return result


def main():
    reader = sys.stdin
    array = list(next(reader).rstrip())
    nr_queries = int(next(reader))
    queries = read_queries(reader, nr_queries)
    result = most_frequent_chars_in_range(array, queries)
    for char in result:
        print(char)


if __name__ == '__main__':
    main()
